presto-hive/src/main/java/com/facebook/presto/hive/util/CloseableAsyncQueue.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@ThreadSafe
+public class CloseableAsyncQueue<T>
+{
+    private static final CompletableFuture<?> COMPLETED_FUTURE_OF_NULL = CompletableFuture.completedFuture(null);
```

Comment:
I named this differently because there's also completed future of empty list in this class.